### DISCUSSION TITLE
表格结构识别：微调模型训练有线表格结构识别模型，效果不好，优化求助

### DISCUSSION THREAD

**Discussion by alice77-zhen** [PaddlePaddle/PaddleOCR#14341] [Q&A]
🔎 Search before asking
 I have searched the PaddleOCR [Docs](https://paddlepaddle.github.io/PaddleOCR/) and found no similar bug report.
 I have searched the PaddleOCR [Issues](https://github.com/PaddlePaddle/PaddleOCR/issues) and found no similar bug report.
 I have searched the PaddleOCR [Discussions](https://github.com/PaddlePaddle/PaddleOCR/discussions) and found no similar bug report.
# 问题描述
使用ch_ppstructure_mobile_v2.0_SLANet_train作为预训练模型，使用20000张表格图片进行训练，所有训练数据都是有线表格。如下所示：
![3](https://github.com/user-attachments/assets/cb69e2f9-b64d-4b50-849b-e426f0e7c4de)
423fb67-9180-42dc-a653-dc0d29599d7b)
模型容易过拟合，训练精度是1时，验证的精度不到70%（训练和验证集的分布是一致的）

# 推理结果如下所示
![image](https://github.com/user-attachments/assets/d2711586-72af-468a-bc7e-80b0fe6d3a48)

# 疑问：
1. 是否需要增加数据量
2. 关于结构的损失函数使用focal loss 会比CrossEntropyLoss的效果更好吗（因为类别不平衡，例如rowspan会比tr/td少）
3. 为什么测试的效果会这么不好呢，还有缺失的部分

此外，还有哪些忽略了的或者还可以优化的方法呢？

**Comment by GreatV** [comment]
针对问题中描述的 **表格结构识别模型微调效果不佳的问题**，结合提供的文档和 PaddleOCR 的相关信息，可以从以下几个方面进行优化和调参尝试：

---

### 1. **数据量与数据多样性**
   - **是否需要增加数据量？**
     - **是的，可能需要增加数据量。** 根据描述，目前使用了 2 万张有线表格图像进行训练。如果模型训练集和验证集分布一致，但仍然出现过拟合现象（训练精度达到 1，但验证精度仅 70%），说明模型可能无法很好地泛化。
     - **如何增加数据量？**
       - 增加更多真实世界的表格样本，尤其是包含不同复杂性、样式和模态（如有线表格、无线表格、合并单元格表格等）的表格。
       - 使用 [TableGeneration](https://github.com/WenmuZhou/TableGeneration) 等工具生成更丰富的模拟表格数据。
       - 对现有数据集进行增强操作，如旋转、缩放、模糊处理，增加模型的鲁棒性。
     - **数据分布多样性** 是提高模型泛化能力的重要因素，尤其是在表格类别不平衡的情况下。

---

### 2. **损失函数优化**
   - **是否需要调整损失函数？**
     - **可以尝试**使用 **Focal Loss** 替代 CrossEntropyLoss。原因是表格数据中不同类别（如 `<td>`、`<tr>`、`rowspan` 等）可能存在类别不平衡问题，Focal Loss 可以聚焦于难分类样本，缓解类别不平衡对模型优化的影响。
     - **如何调整？**
       修改配置文件中损失函数部分的设置，例如：
       ```yaml
       Loss:
         name: FocalLoss
         gamma: 2.0  # 调节因子，可根据实验调整
       ```
     - **实验验证：** 测试不同的 `gamma` 参数值（如 1.0、2.0、3.0），找到最佳损失函数配置。

---

### 3. **模型配置与参数调整**
   - **表格分辨率与 `table_max_len` 参数调整**
     - 针对测试图片分辨率较高的问题，增加 `table_max_len` 是合理的，但需要注意：
       - 增加 `table_max_len` 会显著增加模型的计算复杂度，可能导致推理阶段效果下降。
       - 建议调整图片预处理大小（如缩放到固定宽高）后，再尝试不同的 `table_max_len` 值（如 512、768、1024）。
       - 修改配置文件中的分辨率设置：
         ```yaml
         Global:
           table_max_len: 1024  # 或其他合适的值
           image_shape: [3, 512, 512]  # 根据图片大小调整
         ```
   - **三阶段训练策略**
     - 参考 SLANet 的三阶段训练策略：
       1. **阶段 1：** 使用默认学习率（如 0.001），在原始分辨率（如 488x488）上训练。
       2. **阶段 2：** 加载阶段 1 的最佳模型，降低学习率（如 0.0001）。
       3. **阶段 3：** 加载阶段 2 的最佳模型，调整分辨率（如提升到 512x512），不改变学习率。
     - 通过多阶段训练逐步提升模型性能。

---

### 4. **其他优化方法**
   - **表格结构字典调整**
     - 确保 `character_dict_path` 使用的是适合当前数据的正确表格结构字典。如果训练数据中有自定义标签或特殊标记，需要修改相应的字典文件。
   - **检测与识别模型协同优化**
     - 如果表格检测（如边框检测）或文字识别效果较差，可能会影响表格结构的预测结果。建议检查并优化 `det_model` 和 `rec_model` 的配置。
   - **验证结果的缺失**
     - 如果预测结果中存在缺失部分，可能是模型对某些小单元格或复杂结构的理解能力不足。可以尝试：
       - 增加小尺寸单元格的训练样本。
       - 调整网络结构（如 Backbone），使用更强的特征提取网络（如 ResNet、EfficientNet）。

---

### 5. **性能监控与调试建议**
   - **过拟合问题**
     - 使用正则化方法（如 Dropout 或 Weight Decay）减少过拟合。可以在配置文件中增加正则化参数：
       ```yaml
       Optimizer:
         weight_decay: 0.0001
       ```
   - **模型调试**
     - 可视化模型的中间预测结果（如行列特征、边界框分割）以定位问题。
     - 使用 PaddleOCR 提供的 `tools/debug.py` 检查数据加载、模型预测的具体细节。

---

### 6. **相关参考资源**
   - **PaddleOCR 文档与讨论：**
     - [PaddleOCR 官方文档](https://paddlepaddle.github.io/PaddleOCR/)
     - [PaddleOCR Discussions: 如何提高表格识别精度](https://github.com/PaddlePaddle/PaddleOCR/discussions/14122)
   - **SLANet 模型介绍与配置：**
     - [SLANet 配置文件](https://github.com/PaddlePaddle/PaddleOCR/blob/dygraph/configs/table/SLANet_lcnetv2.yml)
   - **表格生成工具**（可用于生成更多训练数据）：
     - [TableGeneration 项目地址](https://github.com/WenmuZhou/TableGeneration)

---

### 总结
针对当前表格结构识别效果不佳的问题，建议从以下几个方面优化：
1. **增加数据量与多样性**，尤其是复杂表格样本。
2. **调整损失函数**，尝试 Focal Loss 应对类别不平衡。
3. **优化模型参数**，包括 `table_max_len` 和分辨率设置。
4. **采用多阶段训练策略**，逐步提升模型性能。
5. 检查数据预处理、检测模型与字典文件，确保全流程协同优化。

通过以上方法，结合调参实验，相信可以显著提升表格识别模型的精度和鲁棒性！

---
*Response generated by 🤖 [feifei-bot](https://github.com/GreatV/feifei) | chatgpt-4o-latest*

**Reply by alice77-zhen** [reply]
**非常非常感谢您非常快的回答！**

**我还有一些疑问：**
**关于增加数据量的问题：**增加更多真实世界的表格样本，尤其是包含不同复杂性、样式和模态（如有线表格、无线表格、合并单元格表格等）的表格。如果我的场景里只包含有某一种类型的表格，一种增加数据量的方式是增加其他样式的表格，增加泛化性进而增加准确率？另一种方式是只增加这一类型的表格，但是这种方式也可能由于样式少，而增加过拟合的风险，但也提升这个场景的准确率？哪种增加数据的方式才更适合特定类型的表格结构识别呢？

**关于三阶段训练策略：**第一阶段需不需要加载预训练模型（如ch_ppstructure_mobile_v2.0_SLANet_train），如果加载预训练模型，需不需要将学习率调低一点？

**Reply by GreatV** [reply]
试试paddlex的slanet_plus;调调学习率;你用的是不是都是合成数据，加一点真实数据。都要你自己尝试一下的。

**Reply by GreatV** [reply]
三阶段指的是 https://paddlepaddle.github.io/PaddleOCR/latest/algorithm/table_recognition/algorithm_table_slanet.html

**Reply by GreatV** [reply]
上面的回复是机器人回复的，仅供参考。

**Reply by alice77-zhen** [reply]
好的，谢谢。
